libs/ledgerjs/packages/keyring-eth/src/DefaultKeyringEth.ts

```diff
@@ -1,46 +1,120 @@
+  private _contextModule: ContextModule;
 
-  constructor(transport: Transport) {
-    this._appBinding = new AppBinding(transport);
+  constructor(appBinding: AppBinding, contextModule: ContextModule) {
+    this._appBinding = appBinding;
+    this._contextModule = contextModule;
   }
 
   public async signTransaction(
-    _derivationPath: string,
-    _transaction: Transaction,
-    _options: SignTransactionOptions,
+    derivationPath: string,
+    transaction: Transaction,
+    options: SignTransactionOptions,
   ) {
-    // TODO: implement
-    return Promise.resolve({} as EcdsaSignature);
+    const challenge = await this._appBinding.getChallenge();
+
+    const contexts: ContextResponse[] = await this._contextModule.getContexts(transaction, {
+      challenge,
+      options,
+    });
+
+    for (const context of contexts) {
+      if (context.type === "error") {
+        // TODO: handle error here
+        continue;
+      }
+
+      await this._appBinding[context.type](context.payload);
+    }
+
+    const response = await this._appBinding.signTransaction(
+      derivationPath,
+      ethers.utils.serializeTransaction(transaction),
+      null, // context is already fetched by context-module
+    );
+
+    return {
+      r: `0x${response.r}`,
+      s: `0x${response.s}`,
+      v: parseInt(response.v, 16),
+    } as EcdsaSignature;
   }
 
-  public async signMessage(
-    _derivationPath: string,
-    _message: SignMessagePayload,
-    _options: SignMessageOptions,
-  ): Promise<EcdsaSignature> {
-    // TODO: implement
-    return Promise.resolve({} as EcdsaSignature);
+  public async signMessage<Method extends SignMessageMethod>(
+    derivationPath: string,
+    message: SignMessageType<Method>,
+    options: SignMessageOptions<Method>,
+  ) {
+    if (options.method === "personalSign") {
+      const personnalMessage = message as SignMessageType<"personalSign">;
+      const result = await this._appBinding.signPersonalMessage(
+        derivationPath,
+        Buffer.from(personnalMessage).toString("hex"),
+      );
+
+      return result as EcdsaSignature;
+    }
+
+    if (options.method === "eip712") {
+      const eip712Message = message as SignMessageType<"eip712">;
+      try {
+        const result = await this._appBinding.signEIP712Message(derivationPath, eip712Message);
+        return result as EcdsaSignature;
+      } catch (e) {
+        // fallback to eip712Hashed
+        const { EIP712Domain, ...otherTypes } = eip712Message.types;
+        const domainSeparator = ethers.utils._TypedDataEncoder.hashDomain(eip712Message.domain);
+        const hashStruct = ethers.utils._TypedDataEncoder.hashStruct(
+          eip712Message.primaryType,
+          otherTypes,
+          eip712Message.message,
+        );
+
+        const result = await this._appBinding.signEIP712HashedMessage(
+          derivationPath,
+          domainSeparator, // buffer string without 0x
+          hashStruct, // buffer string without 0x
```

Comment:
Ethers isn't providing a `0x${string}` for that ? 😮

Reply:
Yes, ethers is providing a `0x${string}` here